### ISSUE TITLE
GC doesn't print verbose output with OCaml 5

### ISSUE THREAD

**Issue by anmonteiro** [ocaml/ocaml#11692]
The following program is very verbose on 4.14, but produces no output at all in the stderr on OCaml 5:

```ocaml
let () =
  Gc.set { (Gc.get ()) with verbose = 0x017 };
  for _ = 0 to 1000 do
    let s = String.make 1000 'a' in
    print_endline s
  done
```

**Comment by abbysmal** [comment]
Thank you very much for your report: there is indeed a regression here.

The fix is not quite straightforward and requires a bit of decision making.

Currently, this setting is stored in `caml_params`, and not in a global value (like on 4.14's `caml_verb_gc`), nor in `Caml_state`.

The question is, do we want to make it a global variable, or a domain local value? If it is to be a global variable, how much care should we put into making sure domains won't race over it (should Gc.set be a STW section? proper implementations of atomics operations over it (caml_verb_gc? Just domain-local?)

I think a non-atomic, global version of it is ok, but it begs the question of how this function is currently documented and how thread safe it is: for instance, we know it may cause a STW section (when adjusting some GC settings.).
It is thread safe, of course, but the underlying behaviors (STW sections), are currently not documented.

Does anyone has any thoughts on this matter?

**Comment by kayceesrk** [comment]
Since verbose output is for debugging, I would prefer to just have it set globally. Regarding implementation, it would be simplest to make `verb_gc` field in the `caml_params` an `atomic_uintnat` and use relaxed atomics to read and write the value. This should fix the bug.

**Comment by abbysmal** [comment]
Thanks for your answer.

It cannot quite work because `caml_params` cannot be mutated after initialisations.
I decided to revert to the previous approach: having `caml_verb_gc` a global relaxed atomic uintnat.

I have a branch implementing it but I noticed it needs a little bit more work because the verbose values are not quite aligned with what is going on in 4.X.
(for example: `0x1` is the mask for major cycles in 4.X, in 5.0/trunk, it is `0x40`).
We either needs to fix those, or align them to 4.X where possible (and document exceptions)
